all-sdk/external-client-configuration.md

```diff
@@ -0,0 +1,340 @@
+  * This means that camel case parts are separated by underscores _except_ the profile name itself.
+    * So a config value of `profiles.myProfile.tls.clientCertPath` is `TEMPORAL_MYPROFILE_TLS_CLIENT_CERT_PATH`.
+    * 💭 Why?
+      * This is clearer for typing and it is unreasonable to try and break up user-supplied camel case.
+  * Also the profile may contain an underscore and retain that and match that first. So
+    `profiles.my_profile_tls.tls.clientCertPath` is `TEMPORAL_MY_PROFILE_TLS_TLS_CLIENT_CERT_PATH`.
+    * 💭 Why?
+      * It is unreasonable to elide any characters in the user-provided profile name.
+* Environment variables are case-insensitive.
+* The CLI has existing environment variables. Some of these work and some should be still supported but deprecated in
+  favor of the new forms. Below are the CLI env vars and how they apply:
+  * `TEMPORAL_ADDRESS` - works just fine.
+  * `TEMPORAL_NAMESPACE` - works just fine.
+  * `TEMPORAL_API_KEY` - works just fine.
+  * `TEMPORAL_TLS` - works just fine.
+  * `TEMPORAL_TLS_CERT`, `TEMPORAL_TLS_CERT_DATA`, and other mTLS client values - need to be deprecated and replaced
+    with newer forms (only slight changes here).
+    * 💭 Why not just have the `tls` config field be `certData` instead of `clientCertData` so this doesn't have to
+      change?
+      * Not being clear this is for mTLS client use is confusing to users. And we'd have to change the path one anyways
+        because the CLI is not suffixed with path.
+  * `TEMPORAL_TLS_X` for all other values - ones that don't match need to be deprecated and replaced with newer forms.
+  * `TEMPORAL_CODEC_ENDPOINT` - works just fine.
+  * `TEMPORAL_CODEC_AUTH` - works just fine.
+  * 💭 Why deprecate some instead of making all our clients support these variables that exist in CLI?
+    * It is more sane/reasonable to have a simple to understand config-field-to-env-var format than it is to have these
+      special environment variables be inconsistent outliers.
+    * For non-CLI uses (i.e. SDKs), it is unreasonable to burden them with past CLI choices.
+* There are some special environment variables respected by clients:
+  * `TEMPORAL_PROFILE` - the name of the profile to load if one is not provided at load time. The default is  `default`.
+  * `TEMPORAL_CONFIG_FILE` - the path to the config file. The default is `~/.config/temporalio/temporal.json`.
+* ❓ How do to grpc-meta as env var? Like for `My-Header: My-Value`
+  * Env var names are case-insensitive and don't use dashes, so we can't put header name in env var name. And we may not
+    want to canonicalize the header names even though most gRPC clients do it for you.
+  * There's no obvious env var approach to array of strings, and it can't be completely comma-delimited because header
+    values often contain commas. Would rather not force users to have an index in the env var name. Should we just not
+    support gRPC meta from env var until we think this through?
```

Comment:
Could just pass an escaped json object? I'm fine w/ delaying that until later.

Reply:
Yes, that can work but is a bit rough, might be better to just put header name in env var, treat env var underscores as hyphens in header name, and lowercase them all since gRPC and HTTP/2 do that anyways.